File: ChessLab/Engine/DevicePerformance.swift
import Foundation

/// Capacités de l'appareil, pour dimensionner le moteur au plus juste :
///
///  • **Threads sur les cœurs PERFORMANCE** seulement. Les puces Apple sont
///    big.LITTLE ; lancer un thread par cœur logique en met sur les cœurs
///    « efficacité », lents, qui font chauffer l'appareil pour un gain nul
///    (c'était le cas sur iPhone 11 : 4 threads dont 2 sur des cœurs éco).
///  • **Budgets d'analyse par palier** : l'analyse va CHERCHER plus de
///    profondeur sur les appareils modernes (qui l'encaissent dans le temps
///    imparti), et reste sobre en bas de gamme.
///
/// Valeurs lues une fois (elles ne changent pas pendant la session).
enum DevicePerformance {

    /// Nombre de cœurs PERFORMANCE, via `hw.perflevel0` (niveau 0 = les gros
    /// cœurs). Repli sur la moitié des cœurs logiques si l'info manque.
    static let performanceCores: Int = {
        var count: Int32 = 0
        var size = MemoryLayout<Int32>.size
        if sysctlbyname("hw.perflevel0.physicalcpu", &count, &size, nil, 0) == 0, count > 0 {
            return Int(count)
        }
        return max(1, ProcessInfo.processInfo.activeProcessorCount / 2)
    }()

    /// Threads moteur recommandés : les cœurs performance, plafonnés à 4
    /// (au-delà, sur mobile, le gain part en chaleur ; `ThermalMonitor` rabote
    /// encore à 1 en surchauffe).
    static var recommendedThreads: Int { max(1, min(4, performanceCores)) }

    enum Tier { case low, mid, high }

    /// Palier de puissance, indexé sur la RAM (bon proxy de génération) :
    /// ≤4 Go = iPhone 11 / SE, ~6 Go = milieu de gamme récent, ≥8 Go = Pro
    /// récents et iPad M-series.
    static let tier: Tier = {
        let gigabytes = Double(ProcessInfo.processInfo.physicalMemory) / 1_073_741_824
        if gigabytes >= 7.5 { return .high }
        if gigabytes >= 5.5 { return .mid }
        return .low
    }()

    /// Budget de NŒUDS d'une position de classification (hors ouverture).
    ///
    /// Volontairement MODÉRÉ, orienté VITESSE : la revue d'une partie doit
    /// passer tous les coups en ~20-50 s puis s'arrêter (le moteur recompilé
    /// NEON atteint une profondeur suffisante — ~18-20 — dans ce budget, ce
    /// qui suffit amplement à classer les coups et détecter les gaffes). On ne
    /// cherche PAS la profondeur maximale : au-delà, ça consomme sans rien
    /// apporter à la classification (retour utilisateur du 09/08).
    static var classificationNodeBudget: Int {
        switch tier {
        case .low: return 180_000
        case .mid: return 240_000
        case .high: return 300_000
        }
    }

    /// Plafond de TEMPS par position (filet de sécurité quand une position est
    /// dure et que le débit chute). Serré pour garder la passe rapide.
    /// Plafond de la recherche d'AFFINAGE (seconde passe sur un verdict
    /// limite). Plus large que celui de la passe de base : appliquer le même
    /// tronquerait la recherche approfondie au point de la rendre inutile —
    /// on paierait l'attente sans gagner la précision.
    static var refinementCapMs: Int { classificationCapMs * 4 }

    static var classificationCapMs: Int {
        switch tier {
        case .low: return 1_200
        case .mid: return 1_400
        case .high: return 1_600
        }
    }

    /// Profondeur cible de l'analyse en continu (EXPLORATION d'une position
    /// FEN/scan uniquement — jamais en revue de partie). Modérée pour ne pas
    /// faire chauffer inutilement.
    static var liveDepth: Int {
        switch tier {
        case .low: return 18
        case .mid: return 20
        case .high: return 22
        }
    }
}

File: android/engine/src/main/java/com/chesslab/engine/DevicePerformance.kt
package com.chesslab.engine

import android.app.ActivityManager
import android.content.Context

/**
 * Ce que l'appareil peut donner, pour dimensionner le moteur au plus juste.
 * Pendant de `DevicePerformance.swift`.
 *
 * L'analyse va CHERCHER plus de profondeur sur les appareils modernes, qui
 * l'encaissent dans le temps imparti, et reste sobre en bas de gamme. Un
 * budget unique servirait mal les deux : trop lourd d'un côté, trop timide de
 * l'autre.
 *
 * Le palier s'indexe sur la mémoire, bon indicateur de génération. Lu une
 * fois : il ne change pas pendant la session.
 */
object DevicePerformance {

    enum class Tier { low, mid, high }

    @Volatile private var cached: Tier? = null

    fun tier(context: Context): Tier = cached ?: run {
        val am = context.applicationContext.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
        val info = ActivityManager.MemoryInfo().also { am?.getMemoryInfo(it) }
        val gigabytes = info.totalMem / 1_073_741_824.0
        val value = when {
            gigabytes >= 7.5 -> Tier.high
            gigabytes >= 5.5 -> Tier.mid
            else -> Tier.low
        }
        cached = value
        value
    }

    /**
     * Le budget de NŒUDS d'une position de classification, hors ouverture.
     *
     * Volontairement MODÉRÉ, orienté VITESSE : la revue d'une partie doit
     * passer tous les coups en quelques dizaines de secondes puis s'arrêter.
     * On ne cherche PAS la profondeur maximale — au-delà, ça consomme sans
     * rien apporter à la classification.
     */
    fun classificationNodes(context: Context): Long = when (tier(context)) {
        Tier.low -> 180_000
        Tier.mid -> 240_000
        Tier.high -> 300_000
    }

    /**
     * Le plafond de TEMPS par position : un filet de sécurité pour quand une
     * position est dure et que le débit chute. Serré, pour garder la passe
     * rapide.
     */
    fun classificationCapMs(context: Context): Int = when (tier(context)) {
        Tier.low -> 1_200
        Tier.mid -> 1_400
        Tier.high -> 1_600
    }

    /**
     * Le plafond de la recherche d'AFFINAGE. Plus large que celui de la passe
     * de base : appliquer le même tronquerait la recherche approfondie au
     * point de la rendre inutile — on paierait l'attente sans gagner la
     * précision.
     */
    fun refinementCapMs(context: Context): Int = classificationCapMs(context) * 4

    /**
     * La profondeur cible de l'analyse EN CONTINU — exploration d'une position
     * seulement, jamais en revue de partie. Modérée, pour ne pas faire
     * chauffer inutilement.
     */
    fun liveDepth(context: Context): Int = when (tier(context)) {
        Tier.low -> 18
        Tier.mid -> 20
        Tier.high -> 22
    }
}
